scheduler/plugin/plugins.go

```diff
@@ -92,6 +92,11 @@ func NewPluginConfig(pc []v1beta2.PluginConfig) ([]v1beta2.PluginConfig, error)
 	for i := range pc {
 		name := pc[i].Name
 		ret := pluginConfig[name].DeepCopy()
+		// If ret is nil, to reference ret.Object is occurred invalid memory address or nil pointer dereference.
+		// To avoid this error, if ret is nil, we continue to next loop.
+		if ret == nil {
```

Comment:
### [NewPluginConfig](https://github.com/kubernetes-sigs/kube-scheduler-simulator/blob/master/scheduler/plugin/plugins.go)

```go
        for i := range pc {
		name := pc[i].Name
		klog.Info("plugin config: ", name)
		ret := pluginConfig[name].DeepCopy()
		// If ret is nil, to reference ret.Object is occurred invalid memory address or nil pointer dereference.
		// To avoid this error, if ret is nil, we continue to next loop.
		if ret == nil {
			continue
		}

		// v1beta2.PluginConfig may have data in pc[i].Args.Raw as []byte.
		// We have to encoding it in this case.
		if len(pc[i].Args.Raw) != 0 {
			// override default configuration
			if err := json.Unmarshal(pc[i].Args.Raw, &ret.Object); err != nil {
				return nil, err
			}
		}

		if pc[i].Args.Object != nil {
			// If data exists in both PluginConfig.Args.Raw and PluginConfig.Args.Object,
			// PluginConfig.Args.Raw would be ignored
			ret.Object = pc[i].Args.Object
		}

		pluginConfig[name] = ret
	}
```

### Resource

One node and two pods on it, own unscheduled pod.
To reset all resources and reset scheduler configuration, I click reset button.
This behavior outputs as the following.

### output
```console
default plugin config: map[DefaultPreemption:&RawExtension{Raw:nil,} InterPodAffinity:&RawExtension{Raw:nil,} NodeAffinity:&RawExtension{Raw:nil,} NodeResourcesBalancedAllocation:&RawExtension{Raw:nil,} NodeResourcesFit:&RawExtension{Raw:nil,} PodTopologySpread:&RawExtension{Raw:nil,} VolumeBinding:&RawExtension{Raw:nil,}]

/// snip

219 15:41:52.102498       1 plugins.go:95] plugin config: DefaultPreemption

I0219 15:41:52.102583       1 plugins.go:95] plugin config: InterPodAffinity

I0219 15:41:52.102646       1 plugins.go:95] plugin config: NodeAffinity

I0219 15:41:52.102663       1 plugins.go:95] plugin config: NodeResourcesBalancedAllocation

I0219 15:41:52.102725       1 plugins.go:95] plugin config: NodeResourcesFit

I0219 15:41:52.102752       1 plugins.go:95] plugin config: PodTopologySpread

I0219 15:41:52.102770       1 plugins.go:95] plugin config: VolumeBinding

/// snip

default plugin config: map[DefaultPreemption:&RawExtension{Raw:nil,} InterPodAffinity:&RawExtension{Raw:nil,} NodeAffinity:&RawExtension{Raw:nil,} NodeResourcesBalancedAllocation:&RawExtension{Raw:nil,} NodeResourcesFit:&RawExtension{Raw:nil,} PodTopologySpread:&RawExtension{Raw:nil,} VolumeBinding:&RawExtension{Raw:nil,}]

/// snip

I0219 15:42:31.684998       1 plugins.go:95] plugin config: DefaultPreemption

I0219 15:42:31.685039       1 plugins.go:95] plugin config: InterPodAffinity

I0219 15:42:31.685060       1 plugins.go:95] plugin config: NodeAffinity

I0219 15:42:31.685084       1 plugins.go:95] plugin config: NodeResourcesBalancedAllocation

I0219 15:42:31.685114       1 plugins.go:95] plugin config: NodeResourcesFit

I0219 15:42:31.685523       1 plugins.go:95] plugin config: PodTopologySpread

I0219 15:42:31.685557       1 plugins.go:95] plugin config: VolumeBinding

I0219 15:42:31.685607       1 plugins.go:95] plugin config: NodeResourcesBalancedAllocationForSimulator

I0219 15:42:31.685650       1 plugins.go:95] plugin config: InterPodAffinityForSimulator

I0219 15:42:31.685675       1 plugins.go:95] plugin config: NodeResourcesFitForSimulator

I0219 15:42:31.685698       1 plugins.go:95] plugin config: NodeAffinityForSimulator

I0219 15:42:31.685731       1 plugins.go:95] plugin config: PodTopologySpreadForSimulator

I0219 15:42:31.685756       1 plugins.go:95] plugin config: VolumeBindingForSimulator
```

If XXXForSimulator plugin calls DeepCopy(), return value is nil. So, `ret.Object = pc.[i].Args.Object` is occurred invalid memory address or nil pointer dereference.

If this change isn't existed, NewPluginConfig calls panic
```console
goroutine 4069 [running]:

net/http.(*conn).serve.func1(0xc00690fb80)

/usr/local/go/src/net/http/server.go:1804 +0x153

panic(0x2807aa0, 0x43762a0)

/usr/local/go/src/runtime/panic.go:971 +0x499

github.com/kubernetes-sigs/kube-scheduler-simulator/scheduler/plugin.NewPluginConfig(0xc005095880, 0xd, 0x10, 0x0, 0xd, 0x101, 0xc00b7c7c20, 0xc0039c7320)

/go/src/simulator-server/scheduler/plugin/plugins.go:113 +0x2d8

github.com/kubernetes-sigs/kube-scheduler-simulator/scheduler.convertConfigurationForSimulator(0xc0004ce360, 0xc0048c37a0, 0xc006423b80, 0x3063ae8)

/go/src/simulator-server/scheduler/scheduler.go:151 +0x16e

github.com/kubernetes-sigs/kube-scheduler-simulator/scheduler.(*Service).StartScheduler(0xc00643d6e0, 0xc0004ce360, 0x4c11ac, 0x3031910)

/go/src/simulator-server/scheduler/scheduler.go:73 +0x270

github.com/kubernetes-sigs/kube-scheduler-simulator/scheduler.(*Service).RestartScheduler(0xc00643d6e0, 0xc0004ce360, 0x2496bff, 0xc00b745e38)

/go/src/simulator-server/scheduler/scheduler.go:43 +0x6c

github.com/kubernetes-sigs/kube-scheduler-simulator/scheduler.(*Service).ResetScheduler(0xc00643d6e0, 0x0, 0x0)

/go/src/simulator-server/scheduler/scheduler.go:55 +0x34

github.com/kubernetes-sigs/kube-scheduler-simulator/reset.(*Service).Reset(0xc00643d740, 0x3031910, 0xc00647b000, 0x10, 0x10)

/go/src/simulator-server/reset/reset.go:58 +0x28e

github.com/kubernetes-sigs/kube-scheduler-simulator/server/handler.(*ResetHandler).Reset(0xc0052a1c00, 0x306e158, 0xc006562140, 0x1829846755732203, 0xc000111400)

/go/src/simulator-server/server/handler/reset.go:24 +0x8c

github.com/labstack/echo/v4.(*Echo).add.func1(0x306e158, 0xc006562140, 0x2bcc949, 0x20)

/go/pkg/mod/github.com/labstack/echo/v4@v4.5.0/echo.go:544 +0x62

github.com/labstack/echo/v4/middleware.CORSWithConfig.func1.1(0x306e158, 0xc006562140, 0x43a88a0, 0x203002)

/go/pkg/mod/github.com/labstack/echo/v4@v4.5.0/middleware/cors.go:186 +0x1016

github.com/labstack/echo/v4/middleware.LoggerWithConfig.func2.1(0x306e158, 0xc006562140, 0x0, 0x0)

/go/pkg/mod/github.com/labstack/echo/v4@v4.5.0/middleware/logger.go:117 +0x12c

github.com/labstack/echo/v4.(*Echo).ServeHTTP(0xc0074ebb00, 0x302da28, 0xc00c17d960, 0xc0066c9900)

/go/pkg/mod/github.com/labstack/echo/v4@v4.5.0/echo.go:654 +0x2bc

net/http.serverHandler.ServeHTTP(0xc00c49f420, 0x302da28, 0xc00c17d960, 0xc0066c9900)

/usr/local/go/src/net/http/server.go:2867 +0xa3

net/http.(*conn).serve(0xc00690fb80, 0x30319b8, 0xc00647af40)

/usr/local/go/src/net/http/server.go:1932 +0x8cd

created by net/http.(*Server).Serve

/usr/local/go/src/net/http/server.go:2993 +0x39b
```

Reply:
Okay. I think there is another issue. 
Could you change `NewSchedulerService` like this? This field should not be change, but is changing actually because it's pointer and passed without deepcopy.
(It is not needed to revert your change on `NewPluginConfig`.)

```go
// NewSchedulerService starts scheduler and return *Service.
func NewSchedulerService(client clientset.Interface, restclientCfg *restclient.Config, initialSchedulerCfg *v1beta2config.KubeSchedulerConfiguration) *Service {
	iniCfg := initialSchedulerCfg.DeepCopy()
	return &Service{clientset: client, restclientCfg: restclientCfg, initialSchedulerCfg: iniCfg}
}
```

Reply:
👍

I fixed it.